tasking/pulp/tasking/base.py

```diff
@@ -0,0 +1,444 @@
+        :param tags:        A list of tags (strings) to place onto the task, used for searching for
+                            tasks by tag
+        :type  tags:        list
+
+        :param group_id:    The id that identifies which group of tasks a task belongs to
+        :type group_id:     uuid.UUID
+
+        :return:            An AsyncResult instance as returned by Celery's apply_async
+        :rtype:             celery.result.AsyncResult
+        """
+        tag_list = kwargs.pop('tags', [])
+        group_id = kwargs.pop('group_id', None)
+        async_result = super(UserFacingTask, self).apply_async(*args, **kwargs)
+        async_result.tags = tag_list
+
+        # Set the parent attribute if being dispatched inside of a Task
+        parent_arg = self._get_parent_arg
+
+        # Create a new task status with the task id and tags.
+        TaskStatus.objects.create(pk=async_result.id, state=TaskStatus.WAITING, tags=tag_list,
+                                  group=group_id, **parent_arg)
+        return async_result
+
+    def __call__(self, *args, **kwargs):
+        """
+        Update the :class: `pulp.app.models.Task` object, log, and call the task.
+
+        This updates the :attr: `started_at` and sets the :attr: `state` to :attr: `RUNNING`.
+
+        If the :class: `pulp.app.models.Task` status attribute :attr: `state` is
+        :attr: `CANCELED` the task is skipped.
+
+        Skip the status updating if the task is called synchronously.
+        """
+        if not self.request.called_directly:
+            task_status = TaskStatus.objects.get(pk=self.request.id)
+            if task_status.state != TaskStatus.WAITING:
+                msg = _('Task __call__() occurred but Task %s is not at WAITING')
+                _logger.warning(msg % self.request.id)
+            task_status.state = TaskStatus.RUNNING
+            task_status.started_at = timezone.now()
+            task_status.save()
+        _logger.debug("Running task : [%s]" % self.request.id)
+        return super(UserFacingTask, self).__call__(*args, **kwargs)
+
+    def on_success(self, retval, task_id, args, kwargs):
+        """
+        Update the :class: `pulp.app.models.Task` object, log, and save the result.
+
+        This updates the :attr: `finished_at` and sets the :attr: `state` to :attr: `COMPLETED`.
+
+        Skip the status updating if the callback is called synchronously.
+
+        :param retval:  The return value of the task.
+        :type retval:   ???
+
+        :param task_id: Unique id of the executed task.
+        :type task_id:  :class: `uuid.UUID`
+
+        :param args:    Original arguments for the executed task.
+        :type args:     list
+
+        :param kwargs:  Original keyword arguments for the executed task.
+        :type kwargs:   dict
+        """
+        _logger.debug("Task successful : [%s]" % task_id)
+        if not self.request.called_directly:
+            task_status = TaskStatus.objects.get(pk=task_id)
```

Comment:
Same as above. This would be appropriate logic to have on the ``TaskStatus`` model itself. Perhaps you'd call:

``task_status.set_completed(retval)``